Normative: Add Atomics.waitAsync

syg opened the pull request:
- [x] Shipped in Chrome 87 and Safari 16.4.
- [x] HTML [PR](https://github.com/whatwg/html/pull/4613) to implement hooks.
- [x] [test262](https://github.com/tc39/test262/tree/2df6c7d29a18cd33120e791bdbe2043980a893fd/test/built-ins/Atomics/waitAsync)

syg commented on spec.html:
```diff
@@ -879,6 +879,12 @@ <h1>Algorithm Conventions</h1>
 
+    <emu-clause id="sec-parallelism">
+      <h1>Parallelism</h1>
+      <p>A step that ends with “<dfn>in parallel</dfn>” specifies that it is run at the same time as subsequent steps in the algorithm. This specification does not define the precise mechanism by which this is achieved. For example, an implementation may use processes or threads that make use of multiple CPUs or multiple cores of a single CPU. If a parallel step is subdivided, the substeps run se ...
```
This is the main thing I'd like to discuss in call. cc @tc39/ecma262-editors

syg replied:
Probably need to wordsmith agent definitions to allow for parallel steps. In this section I refer to "thread" as the implementation concept to realize parallelism, like HW threads or green threads. In the agent section, I think "execution thread" is a more abstract thing.

The intention is that the agent encompasses a single thread of execution of _observable JS evaluation_. We should probably qualify parallel steps with saying parallel steps are not allowed to affect observable JS evaluation thread. 

Indeed, the parallel step in this proposal shouldn't touch JS values directly. It either wakes up a blocked agent, or it calls out to the host to resolve a promise. In the case of HTML, the host also can't directly resolve promises across agents. What HTML does is it queues up a task on the target agent to perform the resolution. The queuing of the task is the actual parallel operation, the resolution is always done on the agent main execution thread.

syg replied:
> "it is run at the same time as subsequent steps in the algorithm" suggests that once those subsequent steps are done (i.e., when the algorithm returns)

No, that's not what I meant. Any suggestions? "At the same time as other steps"? FWIW I don't have the intuition that "subsequent steps" is bounded by the return in the current algorithm.

jmdyck replied:
>  FWIW I don't have the intuition that "subsequent steps" is bounded by the return in the current algorithm.

Sure, but "subsequent steps **in the algorithm**" is different. Deleting "in the algorithm" will probably help.

syg replied:
Removed parallel steps entirely.

jmdyck commented on spec.html:
```diff
@@ -42406,88 +42519,111 @@ <h1>
+          1. Assert: There is no Waiter Record in _WL_.[[Waiters]] whose [[PromiseCapability]] field is _waiterRecord_.[[PromiseCapability]] and [[AgentSignifier]] field is _waiterRecord_.[[AgentSignifier]].
+          1. Append _waiterRecord_ as the last element of _WL_.[[Waiters]].
+          1. Let _minimumTimeout_ be _waiterRecord_.[[Timeout]].
+          1. If _minimumTimeout_ is finite, then in parallel
```
The definition of "in parallel" distinguishes between (a) "a step that ends with 'in parallel'" and (b) "subsequent steps in the algorithm". Here, the steps 5.a - 5.d come after the step that ends with 'in parallel', and yet they need to be considered part of (a), not (b).

It might be cleaner to put the parallel steps in an abstract closure, and then say something like:
```
Perform _waitAndTimeout_() in parallel.
```
or
```
Spawn _waitAndTimeout_().
```

(And then, instead of `in parallel`, you'd define `Perform ... in parallel` or `Spawn`, and not have to handle "if a parallel step is subdivided".)

syg replied:
Removed parallel steps entirely.

jmdyck commented on spec.html:
```diff
@@ -42406,88 +42519,111 @@ <h1>
           1. Assert: The surrounding agent is in the critical section for _WL_.
-          1. Assert: _W_ is on the list of waiters in _WL_.
-          1. Remove _W_ from the list of waiters in _WL_.
+          1. Assert: _waiterRecord_ is in _WL_.[[Waiters]].
```
```suggestion
          1. Assert: _WL_.[[Waiters]] contains _waiterRecord_.
```

jmdyck commented on spec.html:
```diff
@@ -42406,88 +42519,111 @@ <h1>
-            _minimumTimeout_: a non-negative extended mathematical value,
-          ): a Boolean
+          SuspendThisAgent (
+            _WL_: a WaiterList Record
```
```suggestion
            _WL_: a WaiterList Record,
```

jmdyck commented on spec.html:
```diff
@@ -42497,22 +42633,101 @@ <h1>
         </emu-note>
       </emu-clause>
 
+      <emu-clause id="sec-hostresolveinagent" type="abstract operation">
```
```suggestion
      <emu-clause id="sec-hostresolveinagent" type="host-defined abstract operation">
```

jmdyck commented on spec.html:
```diff
@@ -42497,22 +42633,101 @@ <h1>
+            _index_: an ECMAScript language value,
+            _value_: an ECMAScript language value,
+            _timeout_: an ECMAScript language value,
+          ): either a normal completion containing either an Object or one of *"not-equal"*, *"timed-out"*, or *"ok"*, or a throw completion
```
```suggestion
          ): either a normal completion containing either an Object, *"not-equal"*, *"timed-out"*, or *"ok"*, or a throw completion
```

jmdyck commented on spec.html:
```diff
@@ -42497,22 +42633,101 @@ <h1>
+            1. If _B_ is *false*, throw a *TypeError* exception.
+          1. Let _block_ be _buffer_.[[ArrayBufferData]].
+          1. Let _offset_ be _typedArray_.[[ByteOffset]].
+          1. Let _indexedPosition_ be (_i_ &times; 4) + _offset_.
```
```suggestion
          1. Let _indexedPosition_ be (_i_ × 4) + _offset_.
```

jmdyck commented on spec.html:
```diff
@@ -42497,22 +42633,101 @@ <h1>
+            1. Let _promiseCapability_ be ~blocking~.
+            1. Let _resultObject_ be *undefined*.
+          1. Else,
+            1. Let _promiseCapability_ to ! NewPromiseCapability(%Promise%).
+            1. Let _resultObject_ to OrdinaryObjectCreate(%Object.prototype%).
```
```suggestion
            1. Let _promiseCapability_ be ! NewPromiseCapability(%Promise%).
            1. Let _resultObject_ be OrdinaryObjectCreate(%Object.prototype%).
```

jmdyck commented on spec.html:
```diff
@@ -42497,22 +42633,101 @@ <h1>
+          1. Perform EnterCriticalSection(_WL_).
+          1. Let _elementType_ be TypedArrayElementType(_typedArray_).
+          1. Let _w_ be GetValueFromBuffer(_buffer_, _indexedPosition_, _elementType_, *true*, ~SeqCst~).
+          1. If _v_ is not equal to _w_, then
```
```suggestion
          1. If _v_ ≠ _w_, then
```

jmdyck commented on spec.html:
```diff
@@ -879,6 +879,12 @@ <h1>Algorithm Conventions</h1>
+    <emu-clause id="sec-parallelism">
+      <h1>Parallelism</h1>
+      <p>A step that ends with “<dfn>in parallel</dfn>” specifies that it is run at the same time as subsequent steps in the algorithm. This specification does not define the precise mechanism by which this is achieved. For example, an implementation may use processes or threads that make use of multiple CPUs or multiple cores of a single CPU. If a parallel step is subdivided, the substeps run se ...
+      <p>A parallel step exhibits a data race if it modifies an alias or mutates a specification or ECMAScript language value at the same time as any other step. Parallel steps generally require synchronization, such as critical sections, to avoid data races. This specification does not define the precise mechanism by which synchronization is achieved. For example, an implementation may use locks ...
```
So the editors are saying that, in this spec, no parallel step exhibits a data race (and if they're wrong, the spec is broken in some sense). But the paragraph also talks about how an implementation can avoid data races: why would an implementation need to avoid something that you assert doesn't happen?

Also, you might want to point out that this concept of data race is similar to but distinct from the one defined in "Memory Model".

syg replied:
I'm saying the spec uses synchronization in parallel steps to avoid data races. The waiter stuff currently already has a critical section. If there is a data race, that is a spec bug. The implementation bit is just saying you can implement what the spec just refers to as a "critical section" with a mutex.

syg replied:
Removed parallel steps entirely.

syg commented on spec.html:
```diff
@@ -879,6 +879,12 @@ <h1>Algorithm Conventions</h1>
     <p>Algorithm steps may declare named aliases for any value using the form “Let _x_ be _someValue_”. These aliases are reference-like in that both _x_ and _someValue_ refer to the same underlying data and modifications to either are visible to both. Algorithm steps that want to avoid this reference-like behaviour should explicitly make a copy of the right-hand side: “Let _x_ be a copy of _some ...
     <p>Once declared, an alias may be referenced in any subsequent steps and must not be referenced from steps prior to the alias's declaration. Aliases may be modified using the form “Set _x_ to _someOtherValue_”.</p>
 
+    <emu-clause id="sec-parallelism">
```
Agreed that end of "Runtime Semantics" is better. It doesn't make sense for static algorithm steps to ever be in parallel.

syg commented on spec.html:
```diff
@@ -1045,6 +1045,12 @@ <h1>Implicit Normal Completion</h1>
+
+      <emu-clause id="sec-parallelism">
+        <h1>Parallelism</h1>
+        <p>An abstract operation or Abstract Closure may be invoked <dfn>in parallel</dfn>, which specifies that the operation or closure is run at the same time as subsequent steps. This specification does not define the precise mechanism by which this is achieved. For example, an implementation may use processes or threads that make use of multiple CPUs or multiple cores of a single CPU.</p>
```
Removed parallel steps entirely.

jmdyck commented on spec.html:
```diff
@@ -42300,12 +42315,115 @@ <h1>The Atomics Object</h1>
+              [[Result]]
+            </td>
+            <td>
+              either *"ok"* or *"timed-out"*
```
```suggestion
              *"ok"* or *"timed-out"*
```

syg replied:
Done.

jmdyck commented on spec.html:
```diff
@@ -42300,12 +42315,115 @@ <h1>The Atomics Object</h1>
+        </table>
+      </emu-table>
+      <p>There can be multiple Waiter Records in a WaiterList with the same agent signifier.</p>
+      <p>The agent cluster has a store of WaiterList Records; the store is indexed by (_block_, _i_), where _block_ is a Shared Data Block and _i_ a byte offset into the memory of _block_. WaiterList Records are agent-independent: a lookup in the store of WaiterList Records by (_block_, _i_) will result in the same WaiterList object in any agent in the agent cluster.</p>
```
```suggestion
      <p>The agent cluster has a store of WaiterList Records; the store is indexed by (_block_, _i_), where _block_ is a Shared Data Block and _i_ a byte offset into the memory of _block_. WaiterList Records are agent-independent: a lookup in the store of WaiterList Records by (_block_, _i_) will result in the same WaiterList Record in any agent in the agent cluster.</p>
```

syg replied:
Done.

syg commented on spec.html:
```diff
@@ -12148,6 +12148,21 @@ <h1>Forward Progress</h1>
+    <p>At any time, an ECMAScript implementation may perform the following steps atomically:</p>
+    <emu-alg>
+      1. Let _now_ be the time value (UTC) identifying the current time.
+      1. For each WaiterList _WL_, do
```
...yeah?

syg commented on spec.html:
```diff
@@ -12148,6 +12148,21 @@ <h1>Forward Progress</h1>
+    <emu-alg>
+      1. Let _now_ be the time value (UTC) identifying the current time.
+      1. For each WaiterList _WL_, do
+        1. Perform EnterCriticalSection(_WL_).
```
Why isn't there a surrounding agent? This runs in some agent.

Good call on the "atomically" thing, which seems to be wrong since the critical section already ensures mutual exclusion.

jmdyck replied:
> Why isn't there a surrounding agent? This runs in some agent.

"While an agent's executing thread executes jobs, the agent is the _surrounding agent_ for the code in those jobs." Are these steps being performed as part of a job?

It seems wrong that one agent in one agent cluster would be tweaking WaiterLists in all agent clusters. I would expect that an agent can't affect any agent cluster other than its own.

syg replied:
Oh agent _cluster_. Sure we can qualify it as "intra agent cluster" but that seems definitionally true. WaiterLists are associated with SAB blocks, so how could an agent have gotten WaiterLists from another agent cluster?

This is the same thing as whatever finalization registry is doing, which is admittedly handwavy but I don't really see a problem. I don't think there's any confusion there that the set of non-live objects _S_ is across agents, just like here it's not across agent clusters.

syg replied:
> "While an agent's executing thread executes jobs, the agent is the surrounding agent for the code in those jobs." Are these steps being performed as part of a job?

...yes, sure?

jmdyck replied:
So is the idea something like this?
- Within each agent cluster:
  - an agent is chosen (arbitrarily?), and
  - a job is enqueued for some future time, to:
    - execute these steps with that agent as the surrounding agent;
    - and enqueue another such job.

(And the job-execution constraints guarantee that the agent won't be in any critical section when the job runs?)

syg replied:
It's simpler than that. The idea is:

- Within each agent cluster:
  - choose an agent
    - run some steps in that agent, which can enqueue a Job

These steps are intended to avoid specifying a parallel watchdog thread in spec-ese, so this just says that at any time, an agent can interleave some steps (or just run them, if no other steps are running), which can then enqueue a Job for the host to run to resolve the waitAsync promise. The Job execution guarantees are for the enqueued Job that resolves the promise, not for the code that calls the host to enqueue the Job. The WL critical section takes care of synchronization for calling the host to enqueue the Job.

jmdyck replied:
You lost me there. This algorithm enqueues a Job to resolve a promise?

I was asking about the circumstances under which this algorithm (i.e., the one in "Processing Model of Atomics.waitAsync") would be executed.

syg replied:
> This algorithm enqueues a Job to resolve a promise?

Yes, that's how the host integration works. The notifier agent may differ from the agent that called `Atomics.waitAsync`, and ecma262 cannot reach across agents to directly fulfill another agent's promise. So the host integration works by the notifier agent calling a host hook, which enqueues a Job in the agent that can fulfill the promise.

In the case of timeouts, there is no "notifier agent", so this "Processing Model of Atomics.waitAsync" section is intended to be maximally simple and basically says any agent at any time can check if any async waiter in its agent cluster has had its timeout expired, and if so, call the host hook.

Does that make sense?

jmdyck replied:
> any agent at any time can check if any async waiter in its agent cluster has had its timeout expired, and if so, call the host hook.

Ah, via `NotifyWaiter`, got it.

> The Job execution guarantees are for the enqueued Job that resolves the promise, not for the code that calls the host to enqueue the Job.

(That's assuming that the latter isn't itself a job, which I gather is how you're thinking about it now. But earlier, you indicated it *was* a job, so that's what I based my question on.)

So, updated question: if the algorithm of "Processing Model of Atomics.waitAsync" *isn't* part of a job, then the job-execution guarantees don't apply, so how does it satisfy `EnterCriticalSection`'s assertion that the surrounding agent is not in any critical section?

syg replied:
> So, updated question: if the algorithm of "Processing Model of Atomics.waitAsync" isn't part of a job, then the job-execution guarantees don't apply, so how does it satisfy EnterCriticalSection's assertion that the surrounding agent is not in any critical section?

Yeah, that's true, there's no re-entrancy guard currently.

We can make those steps have job-like guarantees without it making a job, I guess? Another alternative formulation is perhaps to do away with the current host hook entirely and directly encode the the timeout and promise fulfillment into the "processing model". Something like,

```
At any time in an agent _agent_, when the execution context stack is empty, an implementation may
perform the following steps atomically:

1. For each WaiterList _WL_, do
  1. Enter critical section for _WL_.
  1. For each _waiter_ in _WL_ whose [[AgentSignifier]] is _agent_'s signifier, do
    1. If timeout expired, resolve the [[PromiseCapability]]. (We can do this directly since it's the right agent.)
  1. Exit the critical section for _WL_.
```

syg replied:
> But then why not just say it is a job? I.e., is there something about the definition/handling of a job that needs to not apply to this algorithm?

We can. My initial hesitation was from how the HostEnqueueResolveInAgentJob hook was originally formulated. All that job does is to resolve the promise in the right agent, because an agent can only resolve its own promises. So to make the timer watchdog thing _itself_ a job, like HostEnqueueWatchdogTimerJobForSomeWaiter, we'd either need 2 jobs or something more complicated, like a requirement that the host re-enqueue the job if the timer is not yet expired. 

The HostEnqueueWatchdogTimerJobForSomeWaiter thing is also fine by me, and is probably closer to how real implementations work. Now that I've talked through it, I think that version is the cleanest on top of our existing machinery without having to modify much.

> So that 'resolve' would be something like:
Call(_waiter_.[[PromiseCapability]].[[Resolve]], *undefined*, « *"timed-out"* »).
?

Yep.

syg replied:
I refactored the timeout thing to a Job and removed the "processing model" section entirely. PTAL.

syg commented on spec.html:
```diff
@@ -42300,12 +42315,115 @@ <h1>The Atomics Object</h1>
+              a non-negative extended mathematical value
+            </td>
+            <td>
+              The earliest time by which timeout may be triggered.
```
Done.

syg commented on spec.html:
```diff
@@ -12148,6 +12148,21 @@ <h1>Forward Progress</h1>
+      1. Let _now_ be the time value (UTC) identifying the current time.
+      1. For each WaiterList _WL_, do
+        1. Perform EnterCriticalSection(_WL_).
+        1. For each Waiter Record _waiterRecord_ in _WL_.[[Waiters]] such that _now_ ≥ _waiterRecord_.[[TimeoutTime]], do
```
Done.

jmdyck commented on spec.html:
```diff
@@ -11982,12 +11982,12 @@ <h1>InitializeHostDefinedRealm ( ): either a normal completion containing ~unuse
 
-    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, the constituents of an agent belong exclusive ...
-    <p>An agent's executing thread executes a job on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
+    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, Shared Data Blocks, and specification values  ...
```
This insertion is odd. Shared Data Blocks etc *weren't* just listed as constituents of an agent, so it doesn't make sense to say that they're among the constituents that aren't exclusive to the agent.

I think it would be more helpful to expand the Agent Clusters section to say that every Shared Data Block etc 'belongs to' an agent cluster. (Or, an agent cluster comprises zero or more Shared Data Blocks, WaiterList Records, Waiter Records, whatever.)

syg replied:
Good suggestion, I've added a list of spec values to the agent cluster section.

jmdyck commented on spec.html:
```diff
@@ -11982,12 +11982,12 @@ <h1>InitializeHostDefinedRealm ( ): either a normal completion containing ~unuse
-    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, the constituents of an agent belong exclusive ...
-    <p>An agent's executing thread executes a job on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
+    <p>An <dfn id="agent" variants="agents">agent</dfn> comprises a set of ECMAScript execution contexts, an execution context stack, a running execution context, an <dfn id="agent-record" variants="Agent Records">Agent Record</dfn>, and an <dfn id="executing-thread" variants="executing threads">executing thread</dfn>. Except for the executing thread, Shared Data Blocks, and specification values  ...
+    <p>An agent's executing thread executes code on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
```
Could you explain your thinking behind this change and the one below?

I tried writing a comment, but I'm having trouble without knowing your intent.

syg replied:
The intention is to make this section more generic than about jobs. The history wrt jobs IIRC was that we had a very generic job queue scheduler system that no embedder actually used. This was dead code that no host used, so it was very misleading, and we removed it. So when agents talk about "jobs" I'm pretty sure they're talking about this old notion of a job that has been ripped out.

Since then we've redefined a "Job" to be something more specific: an Abstract Closure that's given to the host to schedule and run. So it seems odd to me to say that "surrounding agent" is only well-defined when an agent is executing a Job. The initial script evaluation, for instance, is not part of a Job in the Abstract Closure sense.

So my intention here was to make the notion of agent executing JS code to be more broad than just the restricted notion of Job. I think "executes code" suffices to broad enough. I had originally thought something like "evaluating ECMAScript code", but I want the "surrounding agent" notion to be well-defined in the waitAsync timer stuff, which is arguably not executing ECMAScript code. Thoughts?

jmdyck replied:
Thanks.

> I think "executes code" suffices to broad enough. I had originally thought something like "evaluating ECMAScript code", but I want the "surrounding agent" notion to be well-defined in the waitAsync timer stuff, which is arguably not executing ECMAScript code. Thoughts?

In the spec, the word "code" (outside of "code unit" and "code point") almost always means ECMAScript source code anyway, so "executes code" doesn't have the meaning you want. I think "executes algorithmic steps" would be better. (You could maybe drop the "algorithmic" for subsequent uses.)

```suggestion
    <p>An agent's executing thread executes algorithmic steps on the agent's execution contexts independently of other agents, except that an executing thread may be used as the executing thread by multiple agents, provided none of the agents sharing the thread have an Agent Record whose [[CanBlock]] field is *true*.</p>
```

or maybe "executes algorithms".

----

> The intention is to make this section more generic than about jobs.

I'm okay with the idea that an agent can execute a slightly broader class of thing than just jobs, but I'd like it to be clear that a job is one of the things it can execute. Currently, the only explicit connection between 'agent' and 'job' is in this para and the next, and this PR is dissolving that connection. Maybe somewhere near the definition of job, it could say that a job is executed by an agent. (See also issue #2642.)

> The history wrt jobs IIRC was that we had a very generic job queue scheduler system that no embedder actually used. This was dead code that no host used, so it was very misleading, and we removed it.

Yup, that was PR #1597.

> So when agents talk about "jobs" I'm pretty sure they're talking about this old notion of a job that has been ripped out.

Well, except #1597 also changed these paragraphs, so presumably they were modified to talk about the new notion of a job.

> Since then we've redefined a "Job" to be something more specific: an Abstract Closure that's given to the host to schedule and run.

That was also done in #1597.

> So it seems odd to me to say that "surrounding agent" is only well-defined when an agent is executing a Job. The initial script evaluation, for instance, is not part of a Job in the Abstract Closure sense.

I agree it's odd. (#1597 dropped ScriptEvaluationJob without saying that an agent could execute non-jobs like ScriptEvaluation.)

But if an agent can execute non-jobs, are there any restrictions on when/how it executes them? (E.g., can an agent start executing a ScriptEvaluation while another is in progress?)

syg replied:
I've added a bit to the Job section about being executed by agents.

jmdyck commented on spec.html:
```diff
@@ -11982,12 +11982,12 @@ <h1>InitializeHostDefinedRealm ( ): either a normal completion containing ~unuse
       <p>Some web browsers share a single executing thread across multiple unrelated tabs of a browser window, for example.</p>
     </emu-note>
-    <p>While an agent's executing thread executes jobs, the agent is the <dfn id="surrounding-agent" variants="surrounding agents">surrounding agent</dfn> for the code in those jobs. The code uses the surrounding agent to access the specification-level execution objects held within the agent: the running execution context, the execution context stack, and the Agent Record's fields.</p>
+    <p>While an agent's executing thread is executing code, the agent is the <dfn id="surrounding-agent" variants="surrounding agents">surrounding agent</dfn> for the code being evaluated. The code uses the surrounding agent to access the specification-level execution objects held within the agent: the running execution context, the execution context stack, and the Agent Record's fields.</p>
```
```suggestion
    <p>While an agent's executing thread is executing algorithmic steps, the agent is the <dfn id="surrounding-agent" variants="surrounding agents">surrounding agent</dfn> for those steps. They use the surrounding agent to access the specification-level execution objects held within the agent: the running execution context, the execution context stack, and the Agent Record's fields.</p>
```
(to eliminate anomalous uses of the word "code".)

syg replied:
Done.

jmdyck commented on spec.html:
```diff
@@ -42415,104 +42525,218 @@ <h1>
-          1. Notify the agent _W_.
+          1. Assert: _waiterRecord_.[[Result]] is either *"ok"* or *"timed-out"*.
+          1. If _waiterRecord_.[[PromiseCapability]] is ~blocking~, then
+            1. Wake the agent whose signifier is _waiterRecord_.[[AgentSignifier]] from suspension.
```
Maybe insert a Note something like: "This causes the agent to resume execution in SuspendThisAgent."

syg replied:
Done.

jmdyck commented on spec.html:
```diff
@@ -42415,104 +42525,218 @@ <h1>
-          1. Let _timeout_ be _minimumTimeout_ + _additionalTimeout_.
-          1. NOTE: When _minimumTimeout_ is +∞, _timeout_ is also +∞ and the following step can terminate only by another agent calling NotifyWaiter.
-          1. Perform LeaveCriticalSection(_WL_) and suspend _W_ for up to _timeout_ milliseconds, performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. _W_ can wake from suspension either because the timeout expired or because it was notified explicitly by another agent calling Noti ...
+          1. Perform LeaveCriticalSection(_WL_) and suspend the surrounding agent until the time is _waiterRecord_.[[TimeoutTime]], performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. The surrounding agent can wake from suspension either because the timeout expired or because it w ...
```
The case of "another agent calling NotifyWaiter" now also includes the case where the timeout expired, which maybe calls for some rewording, e.g.:

`The only way for the surrounding agent to wake from suspension is by another agent calling NotifyWaiter with arguments _WL_ and _thisAgent_. This can happen if the timeout expires, or via a call to Atomics.notify.`

syg replied:
Good suggestion.

syg commented on spec.html:
```diff
@@ -42415,104 +42525,218 @@ <h1>
+        <p>Let _timeoutJob_ be a new Job Abstract Closure with no parameters that captures _WL_ and _waiterRecord_ and performs the following steps when called:</p>
+        <emu-alg>
+          1. Perform EnterCriticalSection(_WL_).
+          1. If _WL_.[[Waiters]] does contains _waiterRecord_, then
```
Done.

jmdyck commented on spec.html:
```diff
@@ -42415,104 +42525,219 @@ <h1>
-          1. Assert: _W_ is on the list of waiters in _WL_.
+          1. Assert: _WL_.[[Waiters]] contains _waiterRecord_.
+          1. Let _thisAgent_ be AgentSignifier().
+          1. Assert: _waiterRecord_.[[AgentSignifier]] is _thisAgent_ and _waiterRecord_.[[PromiseCapability]] is ~blocking~.
```
```suggestion
          1. Assert: _waiterRecord_.[[AgentSignifier]] is _thisAgent_.
          1. Assert: _waiterRecord_.[[PromiseCapability]] is ~blocking~.
```

jmdyck commented on spec.html:
```diff
@@ -42415,104 +42525,219 @@ <h1>
+        <h1>
+          DoWait (
+            _mode_: ~sync~ or ~async~,
+            _typedArray_: a TypedArray,
```
```suggestion
            _typedArray_: an ECMAScript language value
```
because the caller isn't ensuring it's a TypedArray.

jmdyck commented on spec.html:
```diff
@@ -42358,28 +42468,28 @@ <h1>
+          1. Assert: The surrounding agent is not in the critical section for any WaiterList Record.
           1. Wait until no agent is in the critical section for _WL_, then enter the critical section for _WL_ (without allowing any other agent to enter).
-          1. If _WL_ has a Synchronize event, then
+          1. If _WL_.[[MostRecentLeaveEvent]] is ~empty~, then
```
```suggestion
          1. If _WL_.[[MostRecentLeaveEvent]] is not ~empty~, then
```

Also, 6 lines down, there's a reference to
`the Synchronize event in _WL_`
which should probably be
`_WL_.[[MostRecentLeaveEvent]]`

ljharb commented on spec.html:
```diff
@@ -42712,30 +42726,41 @@ <h1>
+        <h1>
+          HostEnqueueTimeoutJob (
+            _timeoutJob_: a Job Abstract Closure,
+            _milliseconds_: a Number,
```
should this be a non-negative integer?

syg replied:
Good suggestion, changed it to non-negative integral Number. It's an actual number, not an integer mathematical value, because that feels more correct for the host API: we're not requiring 262 and hosts communicate the reals.

jmdyck commented on spec.html:
```diff
@@ -42415,104 +42524,247 @@ <h1>
+            1. Perform LeaveCriticalSection(_WL_).
+            1. Return ~unused~.
+          1. Let _now_ be the time value (UTC) identifying the current time.
+          1. Perform HostEnqueueTimeoutJob(_timeoutJob_, 𝔽(_waiterRecord_.[[TimeoutTime]]) - _now_).
```
There's nothing to prevent `[[TimeoutTime]]` having a non-integral value, but `HostEnqueueTimeoutJob` requires an integral Number.

Also, add a line `1. Return ~unused~.`

syg replied:
Eh I'll change it back to a Number.

jmdyck commented:
Also, side note: I like how `HostEnqueueCrossAgentJob` and `HostEnqueueTimeoutJob` are now just stubs; all the pseudocode has been hoisted out to the caller. Could we do something similar for `HostEnqueueFinalizationRegistryCleanupJob` (not in this PR)?

syg commented:
> Also, side note: I like how HostEnqueueCrossAgentJob and HostEnqueueTimeoutJob are now just stubs; all the pseudocode has been hoisted out to the caller. Could we do something similar for HostEnqueueFinalizationRegistryCleanupJob (not in this PR)?

Good idea, we should do something similar for FinalizationRegistryCleanupJob.

jmdyck commented:
> Good idea, we should do something similar for FinalizationRegistryCleanupJob.

On second thought, I don't think [HTML's  HostEnqueueFinalizationRegistryCleanupJob](https://html.spec.whatwg.org/#hostenqueuefinalizationregistrycleanupjob) would like that change: it appears to need access to `_finalizationRegistry_`, which it wouldn't have if that was hidden within a Job Abstract Closure.

syg commented on spec.html:
```diff
@@ -42415,104 +42524,247 @@ <h1>
+          1. Return ~unused~.
+        </emu-alg>
+        <p>An implementation of HostEnqueueResolveInAgentJob must schedule _resolveJob_ in the agent signified by _agentSignifier_. The host may delay performing _resolveJob_ in the agent signified by _agentSignifier_, e.g. for resource management reasons, but it eventually must be performed for forward progress.</p>
+        <p>An implementation of HostEnqueueResolveInAgentJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>.</p>
```
Done.

syg commented:
I've rebased this on top of #3058.

Following HTML spec review and with #3058, I've renamed HostEnqueueCrossAgentJob(job, agentSignifier, realm) to HostEnqueueGenericJob(job, realm), simplifying things slightly.

jmdyck commented on spec.html:
```diff
@@ -11976,6 +12002,20 @@ <h1>
+      <h1>
+        HostEnqueueTimeoutJob (
+          _timeoutJob_: a Job Abstract Closure,
+          _milliseconds_: a non-negative Number,
```
I guess "a non-negative Number" includes `+infinity`, but it's unclear if it includes `*NaN*`.

michaelficarra replied:
Yeah I think we should change it to a non-negative finite Number. What's even the point of calling this host hook with +inf? Then we also don't have to worry about the semantics if a NaN finds its way here.

syg replied:
Done.

michaelficarra commented on spec.html:
```diff
@@ -42482,104 +42631,222 @@ <h1>
         <emu-alg>
           1. Assert: The surrounding agent is in the critical section for _WL_.
-          1. Notify the agent _W_.
+          1. Assert: _waiterRecord_.[[Result]] is either *"ok"* or *"timed-out"*.
```
This doesn't narrow the type at all, since those are the only values allowed for `[[Result]]`. I don't think this assert is helpful here.

syg replied:
Removed.

michaelficarra commented on spec.html:
```diff
@@ -42482,104 +42631,222 @@ <h1>
+            1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, « _waiterRecord_.[[Result]] »).
+          1. Else,
+            1. Perform EnqueueResolveInAgentJob(_waiterRecord_.[[AgentSignifier]], _waiterRecord_.[[PromiseCapability]], _waiterRecord_.[[Result]]).
+            1. NOTE: An agent must not access another agent's promise capability in any capacity beyond passing it to the host.
```
I would have probably made this an `emu-note` instead of an inline note, since it applies more generally, at least to these last two branches of the algorithm.

syg replied:
Done.

michaelficarra commented on spec.html:
```diff
@@ -42482,104 +42631,222 @@ <h1>
+          1. Let _timeoutJob_ be a new Job Abstract Closure with no parameters that captures _WL_ and _waiterRecord_ and performs the following steps when called:
+            1. Perform EnterCriticalSection(_WL_).
+            1. If _WL_.[[Waiters]] contains _waiterRecord_, then
+              1. Let _now_ be the time value (UTC) identifying the current time.
```
Let's not name two different variables `now` in the same emu-alg.

syg replied:
Renamed the one inside the Job closure _timeOfJobExecution_.

michaelficarra commented on spec.html:
```diff
@@ -42482,104 +42631,222 @@ <h1>
+          1. Let _q_ be ? ToNumber(_timeout_).
+          1. If _q_ is either *NaN* or *+∞*<sub>𝔽</sub>, let _t_ be +∞; else if _q_ is *-∞*<sub>𝔽</sub>, let _t_ be 0; else let _t_ be max(ℝ(_q_), 0).
+          1. If _mode_ is ~sync~, then
+            1. Let _B_ be AgentCanSuspend().
```
I would inline this.

syg replied:
Done.

michaelficarra reviewed (approved):
LGTM other than some pretty minor comments.

bakkot commented on spec.html:
```diff
@@ -11822,10 +11834,10 @@ <h1>GetGlobalObject ( ): an Object</h1>
 
     <ul>
-      <li>At some future point in time, when there is no running execution context and the execution context stack is empty, the implementation must:
+      <li>At some future point in time in some agent _agent_, when there is no running execution context in _agent_ and its execution context stack is empty, the implementation must:
```
This sounds a bit like we're saying it's generally up to the host which agent to run the job in. That's not what we want, I think - it's always going to be specified.

I think instead

```suggestion
      <li>At some future point in time, when there is no running execution context in the agent for which the job is enqueued and that agent's execution context stack is empty, the implementation must:
```

And then also tweak the paragraph above to make it clear that jobs are scheduled in particular agents:

```
Jobs are scheduled for execution by ECMAScript host environments in a particular agent
```

syg replied:
I think given https://github.com/whatwg/html/pull/4613#discussion_r1193326552, we should probably make all host hooks pass an agent / global.

syg replied:
I've changed the language to say jobs are scheduled to run in a particular realm in a particular agent, and that the hooks take a job and a realm as parameters. HostEnqueueFinalizationRegistryCleanupJob doesn't follow this strictly speaking right now anyway (because it doesn't take a job as an argument, but a `FinalizationRegistry`), which I will fix in a follow up PR.

bakkot commented on spec.html:
```diff
@@ -42482,104 +42631,221 @@ <h1>
+            1. Perform EnterCriticalSection(_WL_).
+            1. If _WL_.[[Waiters]] contains _waiterRecord_, then
+              1. Let _timeOfJobExecution_ be the time value (UTC) identifying the current time.
+              1. Assert: ℝ(_timeOfJobExecution_) ≥ _waiterRecord_.[[TimeoutTime]].
```
This assert isn't guaranteed by anything because time values aren't monotone increasing (because of leap seconds), but we can probably safely ignore that.

syg replied:
Yeah, I'd rather we keep it in? I'd also be happy converting this to a NOTE that the implementation is recommended to use monotone time.

bakkot replied:
Maybe

```suggestion
              1. Assert: ℝ(_timeOfJobExecution_) ≥ _waiterRecord_.[[TimeoutTime]] (assuming that the implementation is using monotone time).
```
or even less formal:

```suggestion
              1. Assert: ℝ(_timeOfJobExecution_) ≥ _waiterRecord_.[[TimeoutTime]] (ignoring potential non-monotonicity of time values).
```

michaelficarra replied:
I like the latter.

syg replied:
Done.

bakkot commented on spec.html:
```diff
@@ -11976,6 +12002,20 @@ <h1>
+        <dt>description</dt>
+        <dd>It schedules _timeoutJob_ to be performed after at least _milliseconds_ milliseconds.</dd>
+      </dl>
+      <p>An implementation of HostEnqueueTimeoutJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>.</p>
```
Hm, actually, is this not required to run in any particular agent? It's up to the host?

If that's so my suggestion above doesn't make sense, but we still should probably reword it to make it clearer that the agent to run the job in is sometimes specified.

bakkot commented on spec.html:
```diff
@@ -42482,104 +42631,221 @@ <h1>
-          1. Let _timeout_ be _minimumTimeout_ + _additionalTimeout_.
-          1. NOTE: When _minimumTimeout_ is +∞, _timeout_ is also +∞ and the following step can terminate only by another agent calling NotifyWaiter.
-          1. Perform LeaveCriticalSection(_WL_) and suspend _W_ for up to _timeout_ milliseconds, performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. _W_ can wake from suspension either because the timeout expired or because it was notified explicitly by another agent calling Noti ...
+          1. Perform LeaveCriticalSection(_WL_) and suspend the surrounding agent until the time is _waiterRecord_.[[TimeoutTime]], performing the combined operation in such a way that a notification that arrives after the critical section is exited but before the suspension takes effect is not lost. The only way for the surrounding agent to wake from suspension is by another agent calling Notify ...
```
If there's only one agent, or if all other agents are busy-waiting, does that imply this can never be woken, even if the timeout expires?

syg replied:
Oops, good catch. For sync suspensions the timeout shouldn't require an extra watchdog agent to trigger the timeout.

bakkot reviewed (commented):
Some comments. Otherwise LGTM.
